sql: revert NaN float comparisons

hueypark opened the pull request:
Fixes: #48020

Release note (sql change): NaN float comparison equals postgres.(NOTE: In order to allow numeric
values to be sorted and used in tree-based indexes, CockroachDB treats NaN values as equal,
and greater than all non-NaN values.)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48211)
<!-- Reviewable:end -->

hueypark commented:
@yuzefovich Should you review this? I haven't succeeded `TestSorterAgainstProcessor`

yuzefovich commented:
I think the issue you're running into is that we might need to update the encoding of NaNs as well - I think we rely on encoding of the values to sort the values correctly, but NaNs are encoded using special markers (see `util/encoding/encoding.go`). That's why Matt added this as a remark to the issue:
```
I haven't done detailed looking to see how this relates to our nan encoding in indexes.
```
To be honest, I'm not very familiar with encodings, so I'm not sure whether NaN encodings do, in fact, need an update. @rohany I think you know some stuff about encodings :)

yuzefovich commented:
Also note that I updated the issue - NaN decimals appear to have the same difference from Postgres.

hueypark commented:
@yuzefovich  Thank you for your help. I will look further.

hueypark commented:
@mjibson I agree that there is a problem with breaking backward compatibility with encoding changes. When I have done, I'll add a detailed description.

hueypark commented:
Even if we didn't think too much, we've already made a decision about it(https://github.com/cockroachdb/cockroach/issues/18860#issuecomment-335933195). I think the same as postgres is good as long as we handle `NaN`, but the decision is up to Cockroach Labs.

We should document the special handling of `NaN` like [postgres](https://www.postgresql.org/docs/current/datatype-numeric.html), even if we do not merge this PR.

> In most implementations of the “not-a-number” concept, NaN is not considered equal to any other numeric value (including NaN). In order to allow numeric values to be sorted and used in tree-based indexes, PostgreSQL treats NaN values as equal, and greater than all non-NaN values.

hueypark commented:
@mjibson I couldn't find a problem with the encoding-related changes. Please let me know if you found it.

yuzefovich reviewed (commented):
I'm concerned about backwards-compatibility. I'll let @jordanlewis share his thoughts on that.

> Reviewed 24 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-:-M6VN_5b6QidUy-U26Ar:bdd9b8t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark and @mjibson)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads.go, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VFGZoEl5V9sTqsOVq:-M6VFGZoEl5V9sTqsOVr:bz51apy) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/colexec/execgen/cmd/execgen/overloads.go#L723)):*
> ```Go
> 			{
> 				a, b := float64({{.Left}}), float64({{.Right}})
> 				{{.Target}} = tree.CompareFloats(a, b)
> ```

~I know this sounds unfortunate, but we want to copy over the code from `CompareFloats` in here (in a sense, to inline it manually).~

Some context: this piece of code is a part of the templates that we're using to generate the code for the vectorized engine, and it will likely to be used in a tight loop. Making function calls is expensive, so - at least at the moment - we're copying over the code that row-by-row engine is using.

Update: I see that we do call `CompareDecimals` already, so it's ok to call `CompareFloats` as well.
___
*[pkg/sql/sem/tree/datum.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VHlaH9-leoVoVmAM5:-M6VHlaH9-leoVoVmAM6:bd2ow6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/sem/tree/datum.go#L811)):*
> ```Go
> 		return 1
> 	}
> 	// NaN sorts before non-NaN (#10109).
> ```

nit: I don't think that this comment is still correct, right?
___
*[pkg/sql/sem/tree/datum.go, line 991 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VHfE37qrM_RJBVZmB:-M6VHfE37qrM_RJBVZmC:b-el5qie) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/sem/tree/datum.go#L991)):*
> ```Go
> 
> // CompareDecimals compares 2 apd.Decimals according to the SQL comparison
> // rules, making sure that NaNs sort first.
> ```

nit: I think now `NaN`s will sort last, right?


<!-- Sent from Reviewable.io -->

mjibson commented on pkg/util/encoding/encoding.go:
```diff
@@ -44,11 +44,12 @@ const (
-	floatZero    = floatNeg + 1
-	floatPos     = floatZero + 1
-	floatNaNDesc = floatPos + 1 // NaN encoded descendingly
+	floatNaNDesc    = encodedNotNull + 1 // NaN encoded descendingly
```
This changes the meaning for values that are already on disk. We essentially cannot change this package. Or rather, if you do change this package, it must meet some properties:

1) Data encoded in the old format must be understandable by any future reader.
2) Data encoded in the new format must be understandable by the previous version, or the new data format will not be written until all nodes have been upgraded to the current version.

This PR violates both of those properties.

mjibson replied:
(I think this is a very tricky issue, and I suggest you consider not working on it due to its annoyance.)

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48211)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark and @yuzefovich)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads.go, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VFGZoEl5V9sTqsOVq:-M6VRi_T9DK7u1xkM6w6:bfgvnhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/colexec/execgen/cmd/execgen/overloads.go#L723)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

~I know this sounds unfortunate, but we want to copy over the code from `CompareFloats` in here (in a sense, to inline it manually).~

Some context: this piece of code is a part of the templates that we're using to generate the code for the vectorized engine, and it will likely to be used in a tight loop. Making function calls is expensive, so - at least at the moment - we're copying over the code that row-by-row engine is using.

Update: I see that we do call `CompareDecimals` already, so it's ok to call `CompareFloats` as well.
</blockquote></details>

Decimals pretty much have to call into apd to work. I'd suggest keeping the code fully inlined here as well and not calling into a function. Floats should be fast.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48211)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark and @mjibson)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads.go, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VFGZoEl5V9sTqsOVq:-M6WTdoW2-O2iKgX6NIY:bxg31x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/colexec/execgen/cmd/execgen/overloads.go#L723)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Decimals pretty much have to call into apd to work. I'd suggest keeping the code fully inlined here as well and not calling into a function. Floats should be fast.
</blockquote></details>

It turns out, the compiler is smart enough to inline the call to `CompareFloats`. Out of curiosity, I ran some benchmarks:
```
name                                                    old time/op    new time/op    delta
ProjOp/projEQFloatFloat/useSel=true/hasNulls=true-16      8.97µs ± 5%    9.00µs ± 2%      ~     (p=0.912 n=10+10)
ProjOp/projEQFloatFloat/useSel=true/hasNulls=false-16     2.26µs ± 3%    8.81µs ± 2%  +289.01%  (p=0.000 n=10+8)
ProjOp/projEQFloatFloat/useSel=false/hasNulls=true-16     7.32µs ± 6%    8.21µs ± 4%   +12.08%  (p=0.000 n=10+10)
ProjOp/projEQFloatFloat/useSel=false/hasNulls=false-16    1.20µs ± 4%    5.96µs ± 4%  +396.05%  (p=0.000 n=10+10)
```
(`old` is this branch, `new` is this branch with addition of `//go:noinline` right above the definition of `tree.CompareFloats`).


<!-- Sent from Reviewable.io -->

hueypark committed:
sql: revert NaN float comparisons

Fixes: #48020

Release note (sql change): NaN float comparison equals postgres.(NOTE: In order to allow numeric
values to be sorted and used in tree-based indexes, CockroachDB treats NaN values as equal,
and greater than all non-NaN values.)

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48211)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @yuzefovich)

___
*[pkg/sql/colexec/execgen/cmd/execgen/overloads.go, line 723 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VFGZoEl5V9sTqsOVq:-M6_vPW_DMffVx1SJSfg:bms0nst) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/colexec/execgen/cmd/execgen/overloads.go#L723)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It turns out, the compiler is smart enough to inline the call to `CompareFloats`. Out of curiosity, I ran some benchmarks:
```
name                                                    old time/op    new time/op    delta
ProjOp/projEQFloatFloat/useSel=true/hasNulls=true-16      8.97µs ± 5%    9.00µs ± 2%      ~     (p=0.912 n=10+10)
ProjOp/projEQFloatFloat/useSel=true/hasNulls=false-16     2.26µs ± 3%    8.81µs ± 2%  +289.01%  (p=0.000 n=10+8)
ProjOp/projEQFloatFloat/useSel=false/hasNulls=true-16     7.32µs ± 6%    8.21µs ± 4%   +12.08%  (p=0.000 n=10+10)
ProjOp/projEQFloatFloat/useSel=false/hasNulls=false-16    1.20µs ± 4%    5.96µs ± 4%  +396.05%  (p=0.000 n=10+10)
```
(`old` is this branch, `new` is this branch with addition of `//go:noinline` right above the definition of `tree.CompareFloats`).
</blockquote></details>

Thank you for your good opinion. I'll keep the code as it is.
___
*[pkg/sql/sem/tree/datum.go, line 811 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VHlaH9-leoVoVmAM5:-M6a-KaZ6KX6EL0BBpTs:bsjy5i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/sem/tree/datum.go#L811)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I don't think that this comment is still correct, right?
</blockquote></details>

I removed comment because it is meaningless now.
___
*[pkg/sql/sem/tree/datum.go, line 991 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6VHfE37qrM_RJBVZmB:-M6a-SIxC8ZXIb3rJC6b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/sql/sem/tree/datum.go#L991)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: I think now `NaN`s will sort last, right?
</blockquote></details>

Done.
___
*[pkg/util/encoding/encoding.go, line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48211#-M6K7B7xwNMPLk-lYZwF-r1-47:-M6eMAm_Aod5vCTt3RXq:b3yewqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c0a177a16dbb84c5b5a4adae3f6c31cc72f9362/pkg/util/encoding/encoding.go#L47)):*
> 1. Data encoded in the old format must be understandable by any future reader.

I solved this problem in a slightly tricky way. Please take a look at this(`decimal.go`'s `DecodeIntoNonsortingDecimal`).

> 2. Data encoded in the new format must be understandable by the previous version, or the new data format will not be written until all nodes have been upgraded to the current version.

It's more tricky... I can not solve this for now. If you have any good ideas, please let me know.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 17 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48211)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark commented:
@yuzefovich Is there a possibility that this will merged? Or should I make this as a draft for anyone who will solve #48020 in the future?

yuzefovich commented:
I feel like we are not eager to merge this change because it fixes an edge case behavior pg-compatibility issue while possibly making crdb-backwards-incompatible change. cc @jordanlewis to share his perspective.

jordanlewis commented:
We can't merge this unless there's a way to make the data on disk backward compatible. If we can't do that, or if there are huge hurdles to jump in order to make that work (as I suspect), I don't think we'll be able to merge this.

hueypark commented:
Thank you for your reply. I made this PR as a draft for a future attempt.
